dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -56,204 +179,256 @@ DeleteExt(HWND hwndDlg, LPCWSTR pszExt)
+    {
+        LPCWSTR exe = GetProgramPath(p);
+        if (!*exe || !QueryFileDescription(exe, p->AppName, _countof(p->AppName)))
+        {
+            p->AppName[0] = UNICODE_NULL;
+            p->AppName[1] = TRUE;
+        }
+    }
+    return p->AppName;
+}
+
+static LPWSTR
+GetTypeName(PFILE_TYPE_ENTRY p, PFILE_TYPE_GLOBALS g)
+{
+    if (!p->FileDescription[1] && !p->FileDescription[0])
+    {
+        p->FileDescription[1] = TRUE;
+        if (p->IsExtension())
+        {
+            SHFILEINFOW fi;
+            if (SHGetFileInfoW(p->FileExtension, 0, &fi, sizeof(fi), SHGFI_TYPENAME |
+                               SHGFI_USEFILEATTRIBUTES) && *fi.szTypeName)
+            {
+                StringCchCopyW(p->FileDescription, _countof(p->FileDescription), fi.szTypeName);
+            }
+            else
+            {
+                // FIXME: Remove this hack when SHGetFileInfo is fixed
+                StringCbPrintfW(p->FileDescription, sizeof(p->FileDescription),
+                                g->DefExtTypeNameFmt, &p->FileExtension[1]);
+            }
+        }
+        else
+        {
+            // FIXME: Fix and use ASSOCSTR_FRIENDLYDOCNAME
+            DWORD cb = sizeof(p->FileDescription), Fallback = TRUE;
+            LPCWSTR ClassKey;
+            HRESULT hr = GetClassKey(*p, ClassKey);
+            HKEY hKey;
+            if (SUCCEEDED(hr) && !RegOpenKeyExW(HKEY_CLASSES_ROOT, ClassKey, 0, KEY_READ, &hKey))
+            {
+                Fallback = RegQueryValueExW(hKey, NULL, 0, NULL, (BYTE*)p->FileDescription, &cb) ||
+                           !*p->FileDescription;
+                RegCloseKey(hKey);
+            }
+            if (Fallback)
+            {
+                StringCchCopyW(p->FileDescription, _countof(p->FileDescription), p->FileExtension);
+            }
+        }
+    }
+    return p->FileDescription;
+}
+
 static void
-SetFileTypeEntryDefaultIcon(PFILE_TYPE_ENTRY Entry)
+InitializeDefaultIcons(PFILE_TYPE_GLOBALS g)
 {
-    Entry->hIconLarge = LoadIconW(shell32_hInstance, MAKEINTRESOURCEW(IDI_SHELL_FOLDER_OPTIONS));
-    INT cxSmall = GetSystemMetrics(SM_CXSMICON);
-    INT cySmall = GetSystemMetrics(SM_CYSMICON);
-    Entry->hIconSmall = HICON(LoadImageW(shell32_hInstance, MAKEINTRESOURCEW(IDI_SHELL_FOLDER_OPTIONS),
-                                         IMAGE_ICON, cxSmall, cySmall, 0));
-    StringCbCopyW(Entry->IconPath, sizeof(Entry->IconPath), g_pszShell32);
-    Entry->nIconIndex = -IDI_SHELL_FOLDER_OPTIONS;
+    const INT ResId = NOASSOCRESID;
+    if (!g->hDefExtIconSmall)
+    {
+        g->hDefExtIconSmall = HICON(LoadImageW(shell32_hInstance, MAKEINTRESOURCEW(ResId),
+                                               IMAGE_ICON, g->IconSize, g->IconSize, 0));
+    }
+
+    if (!ImageList_GetImageCount(g->himlSmall))
+    {
+        int idx = ImageList_AddIcon(g->himlSmall, g->hDefExtIconSmall);
+        ASSERT(idx == 0);
+    }
+}
+
+static PFILE_TYPE_ENTRY
```

Comment:
It was a fancy BOOL that would let you do if ((edittype.pEntry = Normalize(pEntry))) ... but I ended up putting those on separate lines to avoid a nitpick :rofl: